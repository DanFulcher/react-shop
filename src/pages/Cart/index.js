import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Container, Row } from 'react-bootstrap';
import Recipe from '../../components/Recipe';

import './index.scss';

class Cart extends Component {
   handleAddQuantity = (id) => {
     const { addQuantity } = this.props;
     addQuantity(id);
   }

   handleSubQuantity = (id) => {
     const { subQuantity } = this.props;

     subQuantity(id);
   }

   handleRemove = (id) => {
     const { remove } = this.props;

     remove(id);
   }

   render() {
     const {
       items, subShipping, addShipping, total, shipping,
     } = this.props;

     const cart = items.length
       ? (
         items.map(item => (
           <li className="collection__item" key={item.id}>
             <div className="collection__item__img">
               <img src={item.img} alt={item.title} />
             </div>
             <div className="collection__item__desc">
               <h3>{item.title}</h3>
               <p className="price">
                 {`Â£${item.price}`}
               </p>
               <p className="quantity">
                 {`Quantity: ${item.quantity}`}
               </p>
               <div className="arrows">
                 <a
                   href="#"
                   onClick={() => {
                     this.handleAddQuantity(item.id);
                   }}
                 >
                   <i className="arrows__up" />
                 </a>

                 <a
                   href="#"
                   onClick={() => {
                     this.handleSubQuantity(item.id);
                   }}
                 >
                   <i className="arrows__down" />
                 </a>
               </div>
               <button
                 type="button"
                 className="remove"
                 onClick={() => {
                   this.handleRemove(item.id);
                 }}
               />
             </div>

           </li>
         ))
       )
       : (
         <p>Nothing.</p>
       );
     return (
       <section className="cart-page content">
         <Container>
           <Row>
             <h1>Cart</h1>
           </Row>
           <Row>
             <h5>You have ordered:</h5>
           </Row>
           <Row>
             <ul className="collection">
               {cart}
             </ul>
           </Row>
           <Row>
             {items.length >= 1 && (
             <Recipe
               total={total}
               shipping={shipping}
               addShipping={addShipping}
               subShipping={subShipping}
             />
             )}

           </Row>
         </Container>
       </section>
     );
   }
}

Cart.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired,
      img: PropTypes.string,
      desc: PropTypes.string,
    }),
  ),
  addQuantity: PropTypes.func.isRequired,
  subQuantity: PropTypes.func.isRequired,
  remove: PropTypes.func.isRequired,
  total: PropTypes.number,
  shipping: PropTypes.bool,
  addShipping: PropTypes.func.isRequired,
  subShipping: PropTypes.func.isRequired,
};

Cart.defaultProps = {
  items: [],
  total: 0,
  shipping: false,
};

export default Cart;
