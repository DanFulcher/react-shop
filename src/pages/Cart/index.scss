.collection {
    padding-left:0;
    list-style:none;
    width:100%;
    &__item {
        display:flex;
        justify-content: space-between;
        border:1px solid #e3e3e3;
        margin-bottom:20px;
        padding:20px;
        &__img {
            flex-basis:25%;
            max-width:100%;
            img {
                width:100%;
            }
        }
        &__desc {
            flex-basis:70%;
            max-width:100%;
            display:flex;
            flex-direction:column;
            justify-content:center;
            position:relative;
            .arrows {
                display:flex;
                &__up {
                    width: 0; 
                    height: 0; 
                    border-left: 10px solid transparent;
                    border-right: 10px solid transparent;
                    border-bottom: 10px solid #ff7979;
                    margin-right:20px;
                }
                &__down {
                    width: 0; 
                    height: 0; 
                    border-left: 10px solid transparent;
                    border-right: 10px solid transparent;
                    border-top: 10px solid #ff7979;
                }
            }
            .remove {
                position:absolute;
                top:20px;
                right:20px;
                width:45px;
                height:45px;
                background:#ff7979;
                border-radius:50%;
                appearance: none;
                -webkit-appearance: none;
                -moz-appearance: none;
                border:1px solid #ff7979;
                &::before, &::after {
                    content:'';
                    width:50%;
                    height:2px;
                    background:#fff;
                    position: absolute;
                    top:50%;
                    left:50%;
                }
                &::before {
                    transform:translate(-50%, -50%) rotate(-45deg);
                }
                &::after {
                    transform:translate(-50%, -50%) rotate(45deg);
                }
                &:hover {
                    background:#fff;
                    &::before, &::after {
                        background:#ff7979;
                    }
                }
            }
        }
    }
    
}