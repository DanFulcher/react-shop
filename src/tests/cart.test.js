import React from 'react';
import renderer from 'react-test-renderer';
import Cart from '../pages/Cart';

const items = [{
  id: 1, title: 'Winter body', desc: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Minima, ex.', price: 110, img: '../images/item1.jpg',
}];

describe('<Cart /> functions', () => {
  it('tests the Remove From Cart function', () => {
    const removeFromCart = (id) => {
      const [item] = items;
      expect(id).toEqual(item.id);
    };

    const wrapper = renderer.create(<Cart
      items={items}
      remove={removeFromCart}
      addQuantity={() => false}
      subQuantity={() => false}
      addShipping={() => false}
      subShipping={() => false}
    />);

    const inst = wrapper.getInstance();
    expect(inst.render()).toMatchSnapshot();

    inst.handleRemove(1);
  });

  it('tests the add quantity function', () => {
    const addQuantity = (id) => {
      const [item] = items;
      expect(id).toEqual(item.id);
    };

    const wrapper = renderer.create(<Cart
      items={items}
      remove={() => false}
      addQuantity={addQuantity}
      subQuantity={() => false}
      addShipping={() => false}
      subShipping={() => false}
    />);
    const inst = wrapper.getInstance();
    expect(inst.render()).toMatchSnapshot();

    inst.handleAddQuantity(1);
  });

  it('tests the subtract quantity function', () => {
    const subQuantity = (id) => {
      const [item] = items;
      expect(id).toEqual(item.id);
    };

    const wrapper = renderer.create(<Cart
      items={items}
      remove={() => false}
      addQuantity={() => false}
      subQuantity={subQuantity}
      addShipping={() => false}
      subShipping={() => false}
    />);

    const inst = wrapper.getInstance();
    expect(inst.render()).toMatchSnapshot();

    inst.handleSubQuantity(1);
  });
});
