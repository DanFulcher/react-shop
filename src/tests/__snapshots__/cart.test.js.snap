// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> functions tests the Remove From Cart function 1`] = `
<section
  className="cart-page content"
>
  <ForwardRef
    fluid={false}
  >
    <ForwardRef
      noGutters={false}
    >
      <h1>
        Cart
      </h1>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <h5>
        You have ordered:
      </h5>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <ul
        className="collection"
      >
        <li
          className="collection__item"
        >
          <div
            className="collection__item__img"
          >
            <img
              alt="Winter body"
              src="../images/item1.jpg"
            />
          </div>
          <div
            className="collection__item__desc"
          >
            <h3>
              Winter body
            </h3>
            <p
              className="price"
            >
              £110
            </p>
            <p
              className="quantity"
            >
              Quantity: undefined
            </p>
            <div
              className="arrows"
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <i
                  className="arrows__up"
                />
              </a>
              <a
                href="#"
                onClick={[Function]}
              >
                <i
                  className="arrows__down"
                />
              </a>
            </div>
            <button
              className="remove"
              onClick={[Function]}
              type="button"
            />
          </div>
        </li>
      </ul>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <Recipe
        addShipping={[Function]}
        shipping={false}
        subShipping={[Function]}
        total={0}
      />
    </ForwardRef>
  </ForwardRef>
</section>
`;

exports[`<Cart /> functions tests the add quantity function 1`] = `
<section
  className="cart-page content"
>
  <ForwardRef
    fluid={false}
  >
    <ForwardRef
      noGutters={false}
    >
      <h1>
        Cart
      </h1>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <h5>
        You have ordered:
      </h5>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <ul
        className="collection"
      >
        <li
          className="collection__item"
        >
          <div
            className="collection__item__img"
          >
            <img
              alt="Winter body"
              src="../images/item1.jpg"
            />
          </div>
          <div
            className="collection__item__desc"
          >
            <h3>
              Winter body
            </h3>
            <p
              className="price"
            >
              £110
            </p>
            <p
              className="quantity"
            >
              Quantity: undefined
            </p>
            <div
              className="arrows"
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <i
                  className="arrows__up"
                />
              </a>
              <a
                href="#"
                onClick={[Function]}
              >
                <i
                  className="arrows__down"
                />
              </a>
            </div>
            <button
              className="remove"
              onClick={[Function]}
              type="button"
            />
          </div>
        </li>
      </ul>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <Recipe
        addShipping={[Function]}
        shipping={false}
        subShipping={[Function]}
        total={0}
      />
    </ForwardRef>
  </ForwardRef>
</section>
`;

exports[`<Cart /> functions tests the subtract quantity function 1`] = `
<section
  className="cart-page content"
>
  <ForwardRef
    fluid={false}
  >
    <ForwardRef
      noGutters={false}
    >
      <h1>
        Cart
      </h1>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <h5>
        You have ordered:
      </h5>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <ul
        className="collection"
      >
        <li
          className="collection__item"
        >
          <div
            className="collection__item__img"
          >
            <img
              alt="Winter body"
              src="../images/item1.jpg"
            />
          </div>
          <div
            className="collection__item__desc"
          >
            <h3>
              Winter body
            </h3>
            <p
              className="price"
            >
              £110
            </p>
            <p
              className="quantity"
            >
              Quantity: undefined
            </p>
            <div
              className="arrows"
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <i
                  className="arrows__up"
                />
              </a>
              <a
                href="#"
                onClick={[Function]}
              >
                <i
                  className="arrows__down"
                />
              </a>
            </div>
            <button
              className="remove"
              onClick={[Function]}
              type="button"
            />
          </div>
        </li>
      </ul>
    </ForwardRef>
    <ForwardRef
      noGutters={false}
    >
      <Recipe
        addShipping={[Function]}
        shipping={false}
        subShipping={[Function]}
        total={0}
      />
    </ForwardRef>
  </ForwardRef>
</section>
`;
